### PR TITLE
Django caching instrumentation update

### PR DESCRIPTION
**Pull request by antonpirker**
This adds more data to the cache spans and makes adding the cache item size optional. 
This implements parts of following spec https://develop.sentry.dev/sdk/performance/modules/cache/

Documentation for manual instrumentation for the Cache module can be found here:
https://github.com/getsentry/sentry-docs/pull/9926

Load testing of the feature was done here: https://github.com/getsentry/sentry-python/issues/3058

Fixes https://github.com/getsentry/sentry-python/issues/2964

### CODE REVIEW

File: sentry_sdk/integrations/django/caching.py [hunk 1]
```diff
@@ -1,80 +1,138 @@
...
-                span.set_data(SPANDATA.CACHE_HIT, True)
-
-                size = len(str(value))
-                span.set_data(SPANDATA.CACHE_ITEM_SIZE, size)
-
-            else:
-                span.set_data(SPANDATA.CACHE_HIT, False)
+            with capture_internal_exceptions():
+                if address is not None:
+                    # Strip eventually contained username and password in url style location
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:68]: Can you please explain what this comment means? I am struggling to understand it

File: sentry_sdk/integrations/django/caching.py [hunk 2]
```diff
@@ -1,80 +1,138 @@
...
+                            parsed_url.netloc or "",
+                            parsed_url.path or "",
+                        )
+
+                    span.set_data(SPANDATA.NETWORK_PEER_ADDRESS, address)
+
+                if port is not None:
+                    span.set_data(SPANDATA.NETWORK_PEER_PORT, int(port))
+
+                key = _get_key(args, kwargs)
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:82]: Can `key` ever be `None` in the args or kwargs? Asking because `str(None) == "None"`, so just wondering if this is a case we need to handle

File: sentry_sdk/integrations/django/caching.py [hunk 3]
```diff
@@ -1,80 +1,138 @@
...
+                    span.set_data(SPANDATA.CACHE_KEY, key)
+
+                item_size = None
+                if is_get_operation:
+                    if value:
+                        if integration.cache_spans_add_item_size:
+                            item_size = len(str(value))
+                        span.set_data(SPANDATA.CACHE_HIT, True)
+                    else:
+                        span.set_data(SPANDATA.CACHE_HIT, False)
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:93]:
```suggestion
                if is_get_operation:
                    if integration.cache_spans_add_item_size:
                        item_size = len(str(value))
                    span.set_data(SPANDATA.CACHE_HIT, bool(value))  # Or maybe `value is not None` would be better
```

File: sentry_sdk/integrations/django/caching.py [hunk 4]
```diff
@@ -1,80 +1,138 @@
...
+                else:
+                    if integration.cache_spans_add_item_size:
+                        item_size = len(str(args[1]))
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:96]:
```suggestion
                elif integration.cache_spans_add_item_size:
                    item_size = len(str(args[1]))
```

File: sentry_sdk/integrations/django/caching.py [hunk 5]
```diff
@@ -1,80 +1,138 @@
...
 
-def _patch_cache_method(cache, method_name):
-    # type: (CacheHandler, str) -> None
+def _patch_cache_method(cache, method_name, address, port):
+    # type: (CacheHandler, str, Optional[str], Optional[str]) -> None
     from sentry_sdk.integrations.django import DjangoIntegration
 
     original_method = getattr(cache, method_name)
 
     @ensure_integration_enabled(DjangoIntegration, original_method)
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:51]:
```suggestion
```
I think it makes more sense to place this on the `sentry_method` function. It is inaccurate to say that this method maps `original_method` because it has a different signature. Although, perhaps it would be more appropriate to make this change in a separate PR.

File: sentry_sdk/integrations/django/caching.py [hunk 6]
```diff
@@ -1,80 +1,138 @@
...
+                else:
+                    if integration.cache_spans_add_item_size:
+                        item_size = len(str(args[1]))
+
+                if item_size is not None:
+                    span.set_data(SPANDATA.CACHE_ITEM_SIZE, item_size)
 
             return value
 
     @functools.wraps(original_method)
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:103]:
```suggestion
    @ensure_integration_enabled(DjangoIntegration, original_method)
```

To be applied alongside my comment on (original version) line 41; also, likely makes sense to do in separate PR.

File: sentry_sdk/integrations/django/caching.py [hunk 7]
```diff
@@ -1,80 +1,138 @@
...
+            _patch_cache_method(cache, method_name, address, port)
         cache._sentry_patched = True
 
 
+def _get_address_port(settings):
+    # type: (dict[str, Any]) -> Tuple[Optional[str], Optional[str]]
+    address, port = None, None
+    location = settings.get("LOCATION")
+    # TODO: location can also be an array of locations
+    # see: https://docs.djangoproject.com/en/5.0/topics/cache/#redis
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:126]: Are you planning to add code to handle this case in this PR?

File: tests/integrations/django/test_basic.py [hunk 8]
```diff
@@ -1158,240 +1126,3 @@ def dummy(a, b):
...
-            "something",
-            {},
-            "get s",
-        ),  # this should never happen, just making sure that we are not raising an exception in that case.
-    ],
-)
-def test_cache_spans_get_span_description(
-    method_name, args, kwargs, expected_description
-):
-    assert _get_span_description(method_name, args, kwargs) == expected_description
```
Comment (szokeasaurusrex) [inline tests/integrations/django/test_basic.py:1397]: I am assuming all the deleted tests were moved to `test_cache_module.py`, correct?

File: tox.ini [hunk 9]
```diff
@@ -94,7 +94,7 @@ envlist =
     {py3.6,py3.9}-django-v{2.2}
     # - Django 3.x
     {py3.6,py3.9}-django-v{3.0}
-    {py3.6,py3.11}-django-v{3.2}
+    {py3.6,py3.9,py3.11}-django-v{3.2}
```
Comment (szokeasaurusrex) [inline tox.ini:97]: Why add `3.9` here?

**Review by szokeasaurusrex — COMMENTED** [review]
Have some questions and suggestions

File: tests/integrations/django/test_basic.py (hunk 8 above)
Comment (antonpirker) [reply tests/integrations/django/test_basic.py:1397]: yes, exactly.

File: tox.ini (hunk 9 above)
Comment (antonpirker) [reply tox.ini:97]: to improve code coverages coverage reports. we need a python version where tests for Django 3.0 and Django 3.2 are running, else codecov always tells us that one part of a `if django_version < 3.2` in 
`sentry_sdk/integrations/django/caching.py` is not covered by tests.

File: sentry_sdk/integrations/django/caching.py (hunk 7 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:126]: No. For V0 of caches module we ignore cache clusters (at least in Django, not sure about direct redis instrumenation)

File: sentry_sdk/integrations/django/caching.py (hunk 3 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:93]: I will keep my longer version, because it is easier to read

File: sentry_sdk/integrations/django/__init__.py [hunk 10]
```diff
@@ -119,19 +130,23 @@ def __init__(
         signals_spans=True,
         cache_spans=False,
         signals_denylist=None,
+        cache_spans_add_item_size=False,
```
[the remark this answers is not in the document]
Comment (antonpirker) [reply sentry_sdk/integrations/django/__init__.py:133]: The main idea was to reduce overhead out of the box. I will do some load tests with and without item_sizes, and if there is no or not much difference I will remove the option, so give the best experience for the user out of the box.

File: sentry_sdk/integrations/django/caching.py [hunk 11]
```diff
@@ -1,80 +1,138 @@
...
 import sentry_sdk
 from sentry_sdk.consts import OP, SPANDATA
-from sentry_sdk.utils import ensure_integration_enabled
+from sentry_sdk.utils import capture_internal_exceptions, ensure_integration_enabled
 
 
 if TYPE_CHECKING:
     from typing import Any
     from typing import Callable
+    from typing import Tuple
```
[the remark this answers is not in the document]
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:16]: will do!

File: sentry_sdk/integrations/django/caching.py (hunk 1 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:68]: done.

File: sentry_sdk/integrations/django/caching.py (hunk 2 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:83]: I do not think so. The only way it would be `"None"` would be if kwarg["key"] would return None (in _get_key()). And this code  is there since a very long time and no user has ever reported that they see `"None"` as the key.

File: sentry_sdk/integrations/django/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (antonpirker) [reply sentry_sdk/integrations/django/__init__.py]: removed the option, because load testing did not reveal any performance impact

File: sentry_sdk/integrations/django/caching.py (hunk 5 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:62]: Yes, probably better a new PR. Can you please create a ticket @szokeasaurusrex ?

File: sentry_sdk/integrations/django/caching.py (hunk 2 above)
Comment (szokeasaurusrex) [reply sentry_sdk/integrations/django/caching.py:83]: ah okay, did not realize this code was here before

File: sentry_sdk/integrations/django/caching.py (hunk 7 above)
Comment (szokeasaurusrex) [reply sentry_sdk/integrations/django/caching.py:130]: Do we have an issue for adding this feature? If yes, it would be nice if we could link the issue in the `TODO` comment. If there is no issue yet, it might be a good idea to create the issue and link it

File: sentry_sdk/integrations/django/caching.py [hunk 12]
```diff
@@ -1,80 +1,152 @@
...
+        key = args[0]
+        if isinstance(key, dict):
+            # Do not leak sensitive data
+            # `set_many()` has a dict {"key1": "value1", "key2": "value2"} as first argument.
+            # Those values could include sensitive data so we replace them with a placeholder
+            key = {x: SENSITIVE_DATA_SUBSTITUTE for x in key}
+        return str(key)
     elif kwargs is not None and "key" in kwargs:
-        description += str(kwargs["key"])
+        return str(kwargs["key"])
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:42]: Why do we do the sensitive data stripping above but not here?

File: sentry_sdk/integrations/django/caching.py [hunk 13]
```diff
@@ -1,80 +1,152 @@
...
+
+    return ""
+
 
+def _get_span_description(method_name, args, kwargs):
+    # type: (str, Any, Any) -> str
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:48]:
```suggestion
    # type: (str, list[Any], dict[str, Any]) -> str
```

File: sentry_sdk/integrations/django/caching.py [hunk 14]
```diff
@@ -1,80 +1,152 @@
...
 
     original_method = getattr(cache, method_name)
 
     @ensure_integration_enabled(DjangoIntegration, original_method)
-    def _instrument_call(cache, method_name, original_method, args, kwargs):
-        # type: (CacheHandler, str, Callable[..., Any], Any, Any) -> Any
+    def _instrument_call(
+        cache, method_name, original_method, args, kwargs, address, port
+    ):
+        # type: (CacheHandler, str, Callable[..., Any], Any, Any, Optional[str], Optional[int]) -> Any
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:66]:
```suggestion
        # type: (CacheHandler, str, Callable[..., Any], list[Any], dict[str, Any], Optional[str], Optional[int]) -> Any
```

File: sentry_sdk/integrations/django/caching.py [hunk 15]
```diff
@@ -1,80 +1,152 @@
...
+                if port is not None:
+                    span.set_data(SPANDATA.NETWORK_PEER_PORT, port)
+
+                key = _get_key(args, kwargs)
+                if key != "":
+                    span.set_data(SPANDATA.CACHE_KEY, key)
+
+                item_size = None
+                if is_get_operation:
+                    if value:
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:89]:
```suggestion
                    if value is not None:
```
Or, is there a reason why `if value` is more appropriate here?

File: sentry_sdk/integrations/django/caching.py [hunk 16]
```diff
@@ -1,80 +1,152 @@
...
+                        span.set_data(SPANDATA.CACHE_HIT, True)
+                    else:
+                        span.set_data(SPANDATA.CACHE_HIT, False)
+                else:
+                    try:
+                        # 'set' command
+                        item_size = len(str(args[1]))
+                    except IndexError:
+                        # 'set_many' command
+                        item_size = len(str(args[0]))
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:100]: Is there a way to check the command directly?

File: sentry_sdk/integrations/django/caching.py [hunk 17]
```diff
@@ -1,80 +1,152 @@
...
+    # type: (dict[str, Any]) -> tuple[Optional[str], Optional[int]]
+    address, port = None, None
+    location = settings.get("LOCATION")
+    # TODO: location can also be an array of locations
+    # see: https://docs.djangoproject.com/en/5.0/topics/cache/#redis
+    if isinstance(location, str):
+        if ":" in location:
+            address, port = location.rsplit(":", 1)
+        else:
+            address, port = location, None
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:135]: This logic is broken. For example, if `location == "https://example.com"`, then `":" in location` is `True`, so we will end up with `address == "https"` and `port == "//example.com"`.

Perhaps, there is a library that can handle the URL parsing for us? Otherwise, we should use a regular expression here, or at least ensure this case is handled.

Please also add a unit test that ensures that this function handles URLs like https://example.com correctly 🙂

File: sentry_sdk/integrations/django/caching.py [hunk 18]
```diff
@@ -1,80 +1,152 @@
...
+    # remove the username and password from URL to not leak sensitive data.
+    if address is not None and "://" in address:
+        parsed_url = urlparse(address)
+        address = "{}://{}{}".format(
+            parsed_url.scheme or "",
+            parsed_url.netloc or "",
+            parsed_url.path or "",
+        )
+
+    return address, int(port) if port is not None else None
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/django/caching.py:147]: [nitpick] the logic would be clearer to me if we return early with `None, None` when `location` is not a `str`. However, I am also fine with the original code, if you prefer it

```suggestion
def _get_address_port(settings):
    # type: (dict[str, Any]) -> tuple[Optional[str], Optional[int]]
    location = settings.get("LOCATION")
    if not isinstance(location, str):
        return None, None
    
    if ":" in location:  # TODO: Still need to fix this logic
        address, port = location.rsplit(":", 1)
    else:
        address, port = location, None

    # If address is a URL (could also be a string identifier or path to a Unix socket)
    # remove the username and password from URL to not leak sensitive data.
    if "://" in address:
        parsed_url = urlparse(address)
        address = "{}://{}{}".format(
            parsed_url.scheme or "",
            parsed_url.netloc or "",
            parsed_url.path or "",
        )

    return address, int(port) if port is not None else None
```

File: sentry_sdk/integrations/redis/__init__.py [hunk 19]
```diff
@@ -218,6 +218,11 @@ def sentry_patched_execute_command(self, name, *args, **kwargs):
         if data_should_be_truncated:
             description = description[: integration.max_data_size - len("...")] + "..."
 
+        # TODO: Here we could also create the caching spans.
+        # TODO: also need to check if the `name` (if this is the cache key value) matches the prefix we want to configure in __init__ of the integration
+        #       Questions:
+        #       -) We should probablby have the OP.DB_REDIS span and a separate OP.CACHE_GET_ITEM (or set_item) span, right?
+        #       -) We probably need to research what redis commands are used by caching libs.
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/redis/__init__.py:225]: Will these be addressed in this PR? If not, I would suggest linking an issue to each – the questions also should likely only go in the issue, not the comment.

File: sentry_sdk/integrations/redis/__init__.py [hunk 20]
```diff
@@ -358,6 +363,7 @@ class RedisIntegration(Integration):
     def __init__(self, max_data_size=_DEFAULT_MAX_DATA_SIZE):
         # type: (int) -> None
         self.max_data_size = max_data_size
+        # TODO: add some prefix that users can set to specify a cache key
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/redis/__init__.py:366]: Same comment regarding `TODO`

File: sentry_sdk/integrations/redis/asyncio.py [hunk 21]
```diff
@@ -59,6 +59,10 @@ async def _sentry_execute_command(self, name, *args, **kwargs):
 
         description = _get_span_description(name, *args)
 
+        # TODO: Here we could also create the caching spans.
+        #       Questions:
+        #       -) We should probablby have the OP.DB_REDIS span and a separate OP.CACHE_GET_ITEM (or set_item) span, right?
+        #       -) We probably need to research what redis commands are used by caching libs.
```
Comment (szokeasaurusrex) [inline sentry_sdk/integrations/redis/asyncio.py:65]: Same comment regarding `TODO`

File: tests/integrations/django/test_cache_module.py [hunk 22]
```diff
@@ -0,0 +1,562 @@
...
+        "default": {
+            "BACKEND": "django.core.cache.backends.dummy.DummyCache",
+            "LOCATION": [
+                "redis://127.0.0.1:6379",
+                "redis://127.0.0.2:6378",
+                "redis://127.0.0.3:6377",
+            ],
+        }
+    }
+
```
Comment (szokeasaurusrex) [inline tests/integrations/django/test_cache_module.py:81]: Please also add a test with URL lacking a port, e.g. https://example.com, as requested in a previous comment

**Review by szokeasaurusrex — CHANGES_REQUESTED** [review]
I found some logic that appears to be broken – this should be fixed, and we should also add tests for it.

Besides that, I have also added some more minor suggestions and questions.

File: sentry_sdk/integrations/django/caching.py (hunk 15 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:91]: depending on the cache backend and if `get` or `get_many` is called it could be also `[]` `{}`, `""`

File: sentry_sdk/integrations/redis/__init__.py (hunk 19 above)
Comment (antonpirker) [reply sentry_sdk/integrations/redis/__init__.py:225]: done

File: sentry_sdk/integrations/redis/__init__.py (hunk 20 above)
Comment (antonpirker) [reply sentry_sdk/integrations/redis/__init__.py:367]: done.

File: sentry_sdk/integrations/redis/asyncio.py (hunk 21 above)
Comment (antonpirker) [reply sentry_sdk/integrations/redis/asyncio.py:65]: done.

File: sentry_sdk/integrations/django/caching.py (hunk 7 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py]: yes, linked

File: sentry_sdk/integrations/django/caching.py (hunk 17 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py]: Good catch! ⭐

File: sentry_sdk/integrations/django/caching.py (hunk 16 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py:102]: probably. but because `set_many` is more likely a rare edge case (because mostly used when using the low level Django cache api by hand) it is good as it is.

File: sentry_sdk/integrations/django/caching.py (hunk 12 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py]: I refactored the function to always strip sensitive data if it is a dict

File: tests/integrations/django/test_cache_module.py (hunk 22 above)
Comment (antonpirker) [reply tests/integrations/django/test_cache_module.py:91]: done

File: sentry_sdk/integrations/django/caching.py (hunk 18 above)
Comment (antonpirker) [reply sentry_sdk/integrations/django/caching.py]: done

**Review by sentrivana — APPROVED** [review]
